File: csharp/simple-linked-list/SimpleLinkedList.cs
using System.Collections;

public class SimpleLinkedList<T> : IEnumerable<T>
{
    private class Node
    {
        public required T Value { get; set; }

        public Node? Next { get; set; }
    }

    public SimpleLinkedList(params IEnumerable<T> values)
    {
        foreach (var value in values)
        {
            Push(value);
        }
    }

    private Node? root;
    private int count = 0;

    public int Count => count;

    public void Push(T value)
    {
        root = new Node
        {
            Value = value,
            Next = root,
        };
        count++;
    }

    public T Pop()
    {
        if (root is null) throw new InvalidOperationException();

        var value = root.Value;
        root = root?.Next;
        count--;
        return value;
    }

    public IEnumerator<T> GetEnumerator()
    {
        var current = root;

        while (current != null)
        {
            yield return current.Value;
            current = current.Next;
        }
    }    
    
    IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
}

File: cpp/simple-linked-list/simple_linked_list.cpp
#include "simple_linked_list.h"

#include <stdexcept>

namespace simple_linked_list
{

    std::size_t List::size() const
    {
        return current_size;
    }

    void List::push(int entry)
    {
        Element *node = new Element(entry);
        node->next = head;
        head = node;

        ++current_size;
    }

    int List::pop()
    {
        Element *node = head;
        int value = node->data;

        head = head->next;

        delete node;
        --current_size;

        return value;
    }

    void List::reverse()
    {
        Element *prev = nullptr;
        Element *current = head;

        while (current != nullptr)
        {
            Element *next = current->next;

            current->next = prev;
            prev = current;
            current = next;
        }

        head = prev;
    }

    List::~List()
    {
        while (head != nullptr)
        {
            Element *next = head->next;
            delete head;
            head = next;
        }
    }

} // namespace simple_linked_list
